### PR TITLE
Nested Properties and Splatting

### LINKED ISSUE #154: Investigate how Confirm-AzBPResource will handle nested properties
As an engineer, I would like to investigate how Confirm-AzBPResource will handle checking a property key and value if the property is nested deeply in the resource object.

Would we expect to have our own logic to support this? Or would it be better for the user to pass in an OData query type parameter? Could we get [Resource ID](https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/template-functions-resource#return-value-4) to work here somehow?

- [ ] Document investigation and come to agreement with team
- [ ] Create subsequent issues for follow up implementation

### PR DESCRIPTION
**Pull request by msha1026**
- Will handle a scenario where the user can check nested properties
  - One example is checking a VM's OS is Linux. The returned would be the [PSVirtualMachine class](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.commands.compute.models.psvirtualmachine?view=az-ps-latest), and the user would need to check the StorageProfile->OsDisk->OsType property. We would expect the user to call `Confirm-AzBPResource` with the following parameters: `-PropertyKey "StorageProfile.OsDisk.OsType` and `-PropertyValue "Linux"`
- Changes `Confirm-AzBPSqlDatabase` to expect a SQL database name so that we are only checking one database

Closes #154 and #155

### CODE REVIEW

**Comment by marclerwick** [comment]
Yeah, this is pretty much what I was thinking.

When you walk the keys should that be in a try/catch? I'm not sure if an exception would be thrown if the key doesn't exist during the walk.

**Comment by msha1026** [comment]
> Yeah, this is pretty much what I was thinking.
> 
> When you walk the keys should that be in a try/catch? I'm not sure if an exception would be thrown if the key doesn't exist during the walk.

Yea I've tested the scenario if a key doesn't exist and an exception is not thrown. Instead it'll return a null value, which is why I check if the value exists; otherwise, I create another ErrorRecord. Either way, I think introducing a try/catch is a good idea. Merging your changes in now and will introduce try/catch too!

**Comment by marclerwick** [comment]
> > Yeah, this is pretty much what I was thinking.
> > When you walk the keys should that be in a try/catch? I'm not sure if an exception would be thrown if the key doesn't exist during the walk.
> 
> Yea I've tested the scenario if a key doesn't exist and an exception is not thrown. Instead it'll return a null value, which is why I check if the value exists; otherwise, I create another ErrorRecord. Either way, I think introducing a try/catch is a good idea. Merging your changes in now and will introduce try/catch too!

To be clear I was wondering about a scenario in the middle of the walk. For example, if the keys are `x.y.z` when there is no `y` property wouldn't we have a `NullReferenceException` on the next iteration when it tries to access `z` from a null reference in the loop?

**Comment by msha1026** [comment]
> > > Yeah, this is pretty much what I was thinking.
> > > When you walk the keys should that be in a try/catch? I'm not sure if an exception would be thrown if the key doesn't exist during the walk.
> > 
> > 
> > Yea I've tested the scenario if a key doesn't exist and an exception is not thrown. Instead it'll return a null value, which is why I check if the value exists; otherwise, I create another ErrorRecord. Either way, I think introducing a try/catch is a good idea. Merging your changes in now and will introduce try/catch too!
> 
> To be clear I was wondering about a scenario in the middle of the walk. For example, if the keys are `x.y.z` when there is no `y` property wouldn't we have a `NullReferenceException` on the next iteration when it tries to access `z` from a null reference in the loop?

Ah yes. That does not throw an exception either, surprisingly, and it will return null.

**Comment by marclerwick** [comment]
> > > > Yeah, this is pretty much what I was thinking.
> > > > When you walk the keys should that be in a try/catch? I'm not sure if an exception would be thrown if the key doesn't exist during the walk.
> > > 
> > > 
> > > Yea I've tested the scenario if a key doesn't exist and an exception is not thrown. Instead it'll return a null value, which is why I check if the value exists; otherwise, I create another ErrorRecord. Either way, I think introducing a try/catch is a good idea. Merging your changes in now and will introduce try/catch too!
> > 
> > 
> > To be clear I was wondering about a scenario in the middle of the walk. For example, if the keys are `x.y.z` when there is no `y` property wouldn't we have a `NullReferenceException` on the next iteration when it tries to access `z` from a null reference in the loop?
> 
> Ah yes. That does not throw an exception either, surprisingly, and it will return null.

 That's interesting! No need for try/catch then! (at least not there LOL)

File: BenchPress/Helpers/BenchPress.Azure/Common.psm1 [hunk 1]
```diff
@@ -228,12 +254,20 @@ function Confirm-Resource {
...
-        $ConfirmResult.ErrorRecord = `
-          Format-IncorrectValueError -ExpectedKey $PropertyKey `
-                                     -ExpectedValue $PropertyValue `
-                                     -ActualResult $ConfirmResult.ResourceDetails.$PropertyKey
+      $ActualValue = $ConfirmResult.ResourceDetails
+      $Keys = $PropertyKey.Split(".")
+      foreach($Key in $Keys){
+        $ActualValue = $ActualValue.$Key
+      }
+      if ($ActualValue -ne $PropertyValue) {
```
Comment (ozboms) [inline BenchPress/Helpers/BenchPress.Azure/Common.psm1:262]: Is this if-statement supposed to be in the the foreach block to check each of the values?

Comment (marclerwick) [reply BenchPress/Helpers/BenchPress.Azure/Common.psm1:262]: No. The foreach navigates the key path to get to the final property. Only the value of the last property is evaluated.

For example, if the initial "object" is `{"x":{"y":{"z":"finalValue"}}}` and the path were `x.y.z`, then the `$ActualValue` would be the value of property `z` which would be `"finalValue"`. That's the only value that matters for the `if` block.

**Comment by marclerwick** [comment]
@msha1026 I did have one thought though... how would this handle arrays in the path?

**Review by jasonshortphd — APPROVED** [review]

**Comment by msha1026** [comment]
> @msha1026 I did have one thought though... how would this handle arrays in the path?

Hmm that's definitely an interesting scenario that I'll need to mull over on how to approach. To make sure I'm understanding correctly, I pulled an actual property from the PSVirtualMachine class. If a user wanted to check a nested property under [virtual machine extension](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.commands.compute.models.psvirtualmachine.extensions?view=az-ps-latest), we would need to support something like `Extensions[0].InstanceView.Type`?

**Review by ozboms — APPROVED** [review]

**Pull request merged**